library/src/main/java/com/novoda/downloadmanager/lib/BatchQuery.java

```diff
@@ -0,0 +1,244 @@
+                        + "ELSE 2 "
+                        + "END",
+                Downloads.Impl.STATUS_RUNNING,
+                Downloads.Impl.STATUS_PENDING,
+                Downloads.Impl.STATUS_PAUSED_BY_APP,
+                Downloads.Impl.STATUS_BATCH_FAILED,
+                Downloads.Impl.STATUS_SUCCESS
+        );
+
+        private Criteria.Builder criteriaIdBuilder;
+        private Criteria.Builder criteriaStatusBuilder;
+
+        public Builder() {
+            builder = new Criteria.Builder();
+        }
+
+        /**
+         * Filter by batch id
+         *
+         * @param id id of the batch
+         * @return {@link BatchQuery.Builder}
+         */
+        public Builder withId(long id) {
+            this.criteriaIdBuilder = new Criteria.Builder();
+            criteriaIdBuilder
+                    .withSelection(Downloads.Impl.Batches._ID, Criteria.Wildcard.EQUALS)
+                    .withArgument(String.valueOf(id));
+            return this;
+        }
+
+        /**
+         * Sorting in ascending order by sort column
+         *
+         * @param sortColumn sort column
+         * @return {@link BatchQuery.Builder}
+         */
+        public Builder withSortAscendingBy(String sortColumn) {
+            builder.sortBy(sortColumn).ascending();
+            return this;
+        }
+
+        /**
+         * Sorting in descending order by sort column
+         *
+         * @param sortColumn sort column
+         * @return {@link BatchQuery.Builder}
+         */
+        public Builder withSortDescendingBy(String sortColumn) {
+            builder.sortBy(sortColumn).descending();
+            return this;
+        }
+
+        /**
+         * Sorts batches according to the 'liveness' of the download, i.e. in the order:
+         * Downloading, queued, other, paused, failed, completed
+         *
+         * @return {@link BatchQuery.Builder}
+         */
+        public Builder withSortByLiveness() {
+            builder.sortBy(ORDER_BY_LIVENESS);
+            return this;
+        }
+
+        /**
+         * Filter by status
+         *
+         * @param statusFlags status flags that can be combined with "|"
+         *                    one of {@link DownloadManager#STATUS_PAUSED},
+         *                    {@link DownloadManager#STATUS_FAILED},
+         *                    {@link DownloadManager#STATUS_PENDING},
+         *                    {@link DownloadManager#STATUS_SUCCESSFUL},
+         *                    {@link DownloadManager#STATUS_RUNNING}
+         *                    <p/>
+         *                    e.g. withStatusFilter(DownloadManager.STATUS_FAILED | DownloadManager.STATUS_PENDING)
+         * @return {@link BatchQuery.Builder}
+         */
+        public Builder withStatusFilter(@Status int statusFlags) {
+            this.criteriaStatusBuilder = new Criteria.Builder();
+            List<Integer> statuses = buildStatusesFrom(statusFlags);
```

Comment:
Well the bit-wise check is done in `buildStatusesFrom(statusFlags)`. Initially I was doing only the bit-wise check without the for loop, but then there was no way to know when to put and not to put `or` between them (ending with an additional `OR` at the end). Then I moved to not using the flags but an `int[]` for the filtering, but then we loose the nicer looks of the flags + plus the restriction that we can apply with the `@IntDef` annotation (see https://github.com/novoda/download-manager/commit/72265ca176b95473cc3e7fd48612f60c46405151#diff-5fabc9dc335ace9af378b7c1c777a861R43)

Here I'm just getting the list of statuses out of the flags in order to construct the query and to correctly place the `OR` between them. Do you think https://github.com/novoda/download-manager/commit/c7b58607bc6950c146b6d4a0322889762a5543b3 looks better?